[RNMobile] Update "add block" button style in default editor view

SiobhyB opened the pull request:
Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/4692

## Related PRs and Installable Builds

* `Gutenberg Mobile:` https://github.com/wordpress-mobile/gutenberg-mobile/pull/4705
* `Android installable build:` https://github.com/wordpress-mobile/WordPress-Android/pull/16196
* `iOS installable build:` https://github.com/wordpress-mobile/WordPress-iOS/pull/18233

## What?

This PR seeks to make the block inserter more visible in the "default view" of the editor. For the purposes of this PR, "default view" has been interpreted as the view of the editor when no blocks (including the editor's title) are selected.

## Why?

User research indicates that it's not always clear to users what primary actions are available when they first open the editor. By iterating on the default view of the editor, it's hoped that it will be easier for users to discover blocks and, therefore, improve the overall writing/onboarding experience.

## How?

With the changes from this PR, whenever a block _isn't_ selected the block inserter button will be styled more prominently and with an `Add Blocks` label. The following is a high-level overview of the code changes involved:

### Logic around "isDefaultView" check

A [new `noContentSelected` check](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/edit-post/src/components/header/header-toolbar/index.native.js#L178) has been implemented to determine whether a block or title is currently selected or not. If nothing is selected within the editor, then `noContentSelected` returns as `true`.

### Styling
 
The `noContentSelected` check is utilised to add different styles to the toolbar and block inserter button in cases when it returns as `true`:

* The `header-toolbar` component (the toolbar at the bottom of the editor) has been updated to [increase in height]([header-toolbar](https://github.com/WordPress/gutenberg/tree/82eeb5f8aeeca855849649479ea488f955d5cb45/packages/edit-post/src/components/header/header-toolbar)[header-toolbar](https://github.com/WordPress/gutenberg/tree/82eeb5f8aeeca855849649479ea488f955d5cb45/packages/edit-post/src/components/header/header-toolbar)) when `noContentSelected` is `true`. To enable this, a new `$mobile-header-toolbar-expanded-height` variable [has been added](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/base-styles/_variables.scss#L114) to the global `_variables.scss` file.
* The `inserter` component (the `+` button within the toolbar) has also been updated to [display a text label](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/block-editor/src/components/inserter/index.native.js#L55) and [apply new styles](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/block-editor/src/components/inserter/index.native.js#L48) when `noContentSelected` is `true`.

## Testing Instructions

To view and test the new button's new styles, follow these steps within the WordPress Android or iOS app:

* Open the editor and tap the button. 
* By default the title will be selected (this will be changed in a separate PR). In order to view the new button, close the keyboard so that no area of the editor is selected (block or title). 
* The `+` inserter will take up a bigger amount of the toolbar than previously, and will display `Add Blocks` text alongside it.
* Selecting a block within the editor will revert the inserter back to its previous style. 

As this is a UI change, it'd be helpful to test the following as part of the above flow:

* Appearance on dark mode vs. light mode.
* Animation when the inserter button changes from the default style and back.
* Display on different device sizes, including the portrait and landscape modes.
* Appearance with different translations, including longer translations and RTL languages. _Note, no translations have been added as of yet, so testing will need to be done by manually editing code with a local setup._
* Verify the text that's read by VoiceOver (iOS) or Talkback (Android) when the new expanded button style is shown is clear (with no regressions).

## Screenshots or screencast <!-- if applicable -->

### General

These before/after screenshots show the inserter button (the `+` icon to the left of the bottom toolbar) as it displayed before the changes in this PR and after. Note, the styling in the "after" screenshot will only be visible when no part of the editor is selected (no block or title selection):

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/162632365-13b35a32-a410-466c-ba15-33a82e0a4947.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/162632357-00a26760-9f4f-4abc-8efc-5744750af58b.png" width="100%"> |

### Dark mode

| Dark Mode - Default View | Dark Mode - Block View |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/162632027-20b8a9da-a0b4-4170-9ee5-bc78fe7e5e4c.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/162632062-b26c9712-7d0b-4d77-8910-4acd53233cca.png" width="100%"> |

SiobhyB committed:
Update CHANGELOG

SiobhyB committed:
Add 'add content' text

This text will be visible alongside the button when the editor is first opened. It'll be wrapped around conditionals in future commits.

SiobhyB committed:
Add 'isDefaultView' check

The 'isDefaultView' check will be used to specify a style for the button that will only be seen with the editor's default view.

SiobhyB committed:
Rename CSS selectors to follow BEM methodology

SiobhyB committed:
[WIP] Implement styles using 'isDefaultView' check

jhnstn commented on packages/block-editor/src/components/inserter/style.native.scss:
```diff
@@ -1,13 +1,32 @@
 }
 
-.addBlockButtonDark {
+.inserter-menu__add-block-button-icon-dark {
```
minor nitpick: `dark` is a modifier so if we are sticking with BEM naming the selector should be `inserter-menu__add-block-button--dark` so that the modifier is separated by a `--`

SiobhyB replied:
Thanks Jason! I've gone ahead to update this in https://github.com/WordPress/gutenberg/pull/39726/commits/b0693a7c688387407a394b4d026f82a8f03a1686.

SiobhyB committed:
Separate 'defaultViewProps' for readability

With this commit, the props needed for the button's default view styling are separated out into their own 'defaultViewProps' variable. The reason behind this change is that the default props in the main 'ToolbarButton' component had a lot of conditionals, which had become fairly hard to read.

SiobhyB committed:
Improvement to CSS code quality

This commit includes the following tweaks to improve the quality of the PR's CSS:

- In accordance with BEM naming, selectors have been updated to separate any modifiers with a --. e.g. "inserter-menu__add-block-button-dark" is now "inserter-menu__add-block-button--dark".
- The " $white" variable has been used instead of the hex code for the colour white.

SiobhyB committed:
Separate 'defaultViewText' into its own variable

To help with readability, this commit separates 'defaultViewText' into its own variable. It also adds a comment above the 'default view' variables to help with overall clarity.

SiobhyB commented on packages/block-editor/src/components/inserter/index.native.js:
```diff
@@ -32,27 +33,64 @@ import { store as blockEditorStore } from '../../store';
+	onLongPress,
+	isDefaultView,
+} ) => {
+	// The "default view" refers to the editor's appearance when no blocks
```
Interested in thoughts around whether this comment is justified. It felt like the reasoning for separate "default view" variables might not be clear when I was working on this, but perhaps I'm wrong on that.

jhnstn replied:
I like it. "Default view" is a bit vague , which is fine but the comment helps shore up the meaning.

SiobhyB replied:
Great, thank for the confidence check! 🙌

SiobhyB committed:
Update CHANGELOG

SiobhyB committed:
Merge branch 'trunk' into rnmobile/update/default-block-inserter-button

SiobhyB committed:
Remove redundant code causing unit test failure

The unit tests on mobile are currently failing due to the following error:

> TypeError: Cannot read property 'color' of undefined

This commit seeks to resolve that failure by removing the offending, redundant code.

SiobhyB committed:
Add styling for dark mode

SiobhyB committed:
Merge branch 'trunk' into rnmobile/update/default-block-inserter-button

SiobhyB commented:
@jhnstn, @iamthomasbishop, the installable builds for Android ([here](https://github.com/wordpress-mobile/WordPress-Android/pull/16196#issuecomment-1081762187)) and iOS ([here](https://github.com/wordpress-mobile/WordPress-iOS/pull/18233#issuecomment-1081738076)) are up and the code's ready for review now. :) 

In addition to the pending questions at https://github.com/wordpress-mobile/gutenberg-mobile/issues/4692#issuecomment-1077829264, I wanted to note that I anticipate there will probably be design changes needed to the button's padding/spacing/look. Looking forward to hearing any feedback!

jhnstn commented:
I ran through the test steps on the iOS build and ran the branch locally on Android. 
Works as expected 🎉 

Looks like we don't have a translation yet for `add content` ( or at least not one for the language I tried, German) so I tried changing the text directly to something longer. Randomly tried Hungarian and noticed it's quite long according to Google translate : `Tartalom hozzáadása lehetőségre` . That length of text looks fine on a Pixel 6
![Screenshot_20220329-170803](https://user-images.githubusercontent.com/744755/160708572-65656c4c-8418-4a2b-9225-5b857ad669ff.png)

I then tried on a smaller device (Nexus One emulator) and noticed the block navigation gets pushed off the screen: 

![Screenshot_20220329_172442](https://user-images.githubusercontent.com/744755/160710496-819a25e8-121e-475c-bc1b-562dc6e6b699.png)

This feels like an edge case but we might want to set a fixed width of the add content buttons and have the text wrap so that the block navigation is always visible.

I also just noticed from the screenshots that the vertical text alignment seems off with the Hungarian text. I'm wondering if that has to do with the accent marks ?

SiobhyB committed:
Vertically align text within button

The text within the "add block" button was previously aligned via padding, which didn't guarantee consistent alignment across devices. With this commit, the padding has been replaced with "align-self: center", which will guarantee the text is always vertical centred.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/update/default-block-inserter-button

SiobhyB commented:
Thanks for this @jhnstn! I hadn't thought of testing longer translations. 🙇‍♀️ 

> This feels like an edge case but we might want to set a fixed width of the add content buttons and have the text wrap so that the block navigation is always visible.

This might be tricky as the toolbar has a [set height of 44px here](https://github.com/WordPress/gutenberg/blob/30f900ec5e925bdd218a8845db533765834e9e0b/packages/edit-post/src/components/header/header-toolbar/style.native.scss#L3), which would need to be changed in order to have the text wrap onto a new line. I'm also wondering whether this is necessarily a blocker as the undo/redo buttons are still accessible via scrolling. If the user's on a smaller device, I'm imagining they'd already have a need to scroll through the toolbar more than users with bigger devices and it may already be part of their flow. That said, I agree it'd be ideal if all the buttons were available without scrolling and am curious to hear a design perspective.

As an aside, the undo/redo buttons are inactive when first opening a post, which would arguably be when the button is most on display. Perhaps there's a case for hiding them from view when inactive?

> I also just noticed from the screenshots that the vertical text alignment seems off with the Hungarian text. I'm wondering if that has to do with the accent marks?

I think this was the result of me using `padding-top` to align the text instead of a more consistent, accurate way. 🤦‍♀️ I've gone ahead to replace `padding-top` with `align-self` in https://github.com/WordPress/gutenberg/pull/39726/commits/38df61c4dd201b39a860f659f8de74720a5b537f to address that.

SiobhyB committed:
Update text for clarity

Recent design feedback suggested that "Add content" was a bit confusing in conjunction with the editor's "Start writing" placeholder. As such, this commit updates that text to the suggested "Add block" alternative.

jhnstn reviewed (approved):
Everything looks good and I didn't find anything while testing. Not sure if we are awaiting any design feedback but I have nothing more to add. Nice work!

SiobhyB commented:
Yay, thanks @jhnstn! 🎉 I'll wait for @iamthomasbishop to look this over before merging anything, too.

iamthomasbishop commented:
@SiobhyB I haven't had a chance to do a thorough review yet, but I noticed one thing at first scan. Our original designs had the height of the "default"/resting state slightly taller to accommodate a slightly larger 'add content' button — I believe it was 56px, whereas the existing toolbar is 44px. Example:

<a href="https://cldup.com/rX6p6VeSLD.png"><img src="https://cldup.com/rX6p6VeSLD.png" width="600"></a>

SiobhyB committed:
Assign larger height to toolbar in default view

With this PR, a new .scss variable is introduced to increase the height of the toolbar for the default editor view. This required "lifting" the existing "isDefaultView" value from the inserter component to the header-toolbar component.

SiobhyB committed:
Hide new default view style when title's selected

SiobhyB committed:
Merge branch 'trunk' into rnmobile/update/default-block-inserter-button

SiobhyB committed:
Apply CSS for default toolbar

This commit fixes an issue with the toolbar's styling, in which the default styles were not being applied as they were applied *before* the main styles. This commit addresses that by applying the default styles *after* the main ones.

SiobhyB committed:
Tweaks to button's styling

iamthomasbishop commented:
@SiobhyB I went through and fine-tuned the sizing/spacing on the button (and as a result, the toolbar). Here's what we're looking at for specs:

## Button States

### Exanded

![](https://cldup.com/gIGJkJgZHb.png)

### Compact

![](https://cldup.com/eF5I6vNK4x.png)

## Prototype

I put together a quick prototype to show what would look like in practice:

https://user-images.githubusercontent.com/1200432/162085726-eec9ecff-d0d2-4c06-80ca-e3cf7d24693a.mp4

SiobhyB committed:
Update text in expanded button

SiobhyB committed:
Set icon size to 24 (the default)

SiobhyB committed:
Update height of editor's default toolbar

SiobhyB committed:
Refine expanded button's padding

SiobhyB commented:
Thanks so much for this, @iamthomasbishop!

I've gone ahead to add the specific values for height and padding that you've set out. I believe that the values around the fonts (size and family) are the defaults, but please let me know if I'm mistaken on that.

For the background colour: `$blue-wordpress` is currently in use as the existing "collapsed" icon's background-colour, which is what's being used for the expanded version too. Should I update that to `$blue-50` for both the "collapsed" and "expanded" versions?

iamthomasbishop commented:
> Should I update that to $blue-50 for both the "collapsed" and "expanded" versions?

That'd be great. It just uses the default accent/blue which should be blue-50 in light mode, blue-30 in dark mode.

SiobhyB committed:
Update button's background colour in light mode

iamthomasbishop commented:
@SiobhyB I just noticed something while testing the latest iOS build, apologies if this is a known issue. If I tap on the compact inserter button (i.e. while the keyboard is shown) I'm seeing the button transform momentarily to the expanded state as the inserter bottomSheet enters. Here's an example:

https://user-images.githubusercontent.com/1200432/162212177-a329da23-2a3d-4cee-902b-b3c96f6dbf2c.MOV

I wonder if this is made worse because the transition on the button states is immediate, but in either way I wouldn't expect the state to change in this specific case.

SiobhyB commented:
@iamthomasbishop, ah, not a known issue, thank you for spotting that! I'll take a look.

SiobhyB committed:
Prevent expanded button displaying in some cases

As reported here, the new expanded button was displaying in some cases when the inserter button was tapped: https://github.com/WordPress/gutenberg/pull/39726#issuecomment-1091752804

This was happening when the default block was selected. This is because the default block is the only block type that's replaced (and therefore de-selected) as part of the inserter flow.

With this commit, we now check for whether the default block is being replaced (if so, it's value will be "null"). If it is, the value of "isDefaultView" is set to "false".

SiobhyB committed:
Merge branch 'trunk' into rnmobile/update/default-block-inserter-button

hypest commented:
👋 @SiobhyB ! I gave https://github.com/WordPress/gutenberg/pull/39726/commits/82eeb5f8aeeca855849649479ea488f955d5cb45 a try and looks like I don't see the text on the Inserter button at all, on the demo app. I tried and reverted https://github.com/WordPress/gutenberg/pull/39726/commits/d2822fe0bf8467a6f3b2dbef8b7cfdc1be80abb6 to see if that could be the culprit and indeed, it looks like that. See the screencast below where I do the revert and then undoing the revert:

https://user-images.githubusercontent.com/1032913/162402460-022848e6-cc85-47a8-988e-e8bb379de8fe.mov

SiobhyB committed:
Revert "false fix"

This commit reverts the code from the previous commit, which turned out to be a "false fix". It didn't resolve the bug it was attempting to address.

hypest commented on packages/block-editor/src/components/inserter/index.native.js:
```diff
@@ -32,27 +33,51 @@ import { store as blockEditorStore } from '../../store';
+			isDisabled={ disabled }
+			{ ...defaultViewProps }
+		>
+			{ isDefaultView && defaultViewText }
```
👋 @SiobhyB , not sure if you have tried this already for fixing the button flashing issue, but wanted to share some of my findings: a bit of promising success by using the `shouldReplaceBlock` state variable in the Inserter class. Specifically, by adding a `&& ! shouldReplaceBlock` conditional. This way, the text label is not shown when the editor is in the process of replacing the default richtext block.

But as I said, that's half-success, since the toolbar styling is not correct, since that's governed by the `HeaderToolbar` component. Not sure yet how to pass such transitional state to the HeaderToolbar yet.

Hope this all helps!

SiobhyB replied:
@hypest, thank you so much for this! 🙇‍♀️ 

I think you're right that the trickiest part to this is with updating the parent component. With that in mind, I've made a change to set `isDefaultView` as an editor setting, which I envisage as being similar to other editor settings that influence the editor's appearance, such as `fixedToolbar`, `reducedUI`, `focusMode`, etc. 

I've been able to get some code that's _functionally_ working with that approach (which I've committed to this PR) but it still has some issues. The biggest of which is: As my latest code changes involve an import of `@wordpress/editor` inside a `@wordpress/block-editor` file, the automated tests fail with the `Cannot read property 'SETTINGS_DEFAULTS' of undefined, js engine: hermes` error (same issue that ended up blocking this PR: https://github.com/WordPress/gutenberg/pull/35021).

I feel like this is on the right track, though, and it's a case of finding a better way to control the new setting's data. I'll keep looking into this, but want to be transparent that I don't think it's something that will be quick for me to get to the bottom of.

hypest replied:
> With that in mind, I've made a change to set isDefaultView as an editor setting, which I envisage as being similar to other editor settings that influence the editor's appearance, such as fixedToolbar, reducedUI, focusMode, etc

Hmm, feels to me that the this showing/hiding of the textual part of the inserter button is not a setting per se. I mean, it's not configuration that we are expecting to hold through the lifecycle of the editor instance. Instead, it's state that changes depending on the user/caret focus. Are there other similar examples already in place perhaps, where we use the settings for state like this?

Perhaps what we need to "copy" is the way the settings are implemented (hooks/events/etc) but not piggyback on the settings component itself.

That said, I wonder if @jhnstn you have thoughts on this?

SiobhyB replied:
> Instead, it's state that changes depending on the user/caret focus. Are there other similar examples already in place perhaps, where we use the settings for state like this?

The only other one that comes to mind is that we use the settings to update `impressions`, which is then used to update parts of the native UI. It was seeing [this line of code](https://github.com/WordPress/gutenberg/blob/95cda71f51ccdb2d764d2c142e746517bdd5a2b6/packages/block-editor/src/components/inserter/index.native.js#L210) in the Inserter component that got me thinking of using the approach I've taken in the latest commits. It feels kind of similar to me in that the `impressions` are more of a state than a setting.

> Perhaps what we need to "copy" is the way the settings are implemented (hooks/events/etc) but not piggyback on the settings component itself.

I wonder if there might be a way to use `context` to change this data. 🤔  Looking forward to hearing any ideas from Jason!

fluiddot replied:
This issue was addressed in https://github.com/WordPress/gutenberg/pull/39726/commits/bc9b4ea72674bbe43d5070deb0239830897374c9.

SiobhyB committed:
Turn 'isDefaultView' into 'global' editor setting

It has become necessary to change the value of 'isDefaultView' at various levels of components, which is difficult to do with the existing implementation (i.e. it's currently defined in a parent component, making it difficult to update its value from within a child component).

With this commit, 'isDefaultView' is converted into a block editor setting, similar to other settings like 'hasFixedToolbar', 'focusMode', 'hasReducedUI', etc.

Following this refactor, it will be easier to change the value of 'isDefaultView' at various component levels. It will also make it easier to iterate on the usage of the value in the future.

SiobhyB committed:
Prevent unwanted 'flash' of 'expanded' button

It's necessary to manually toggle 'isDefaultView' off to prevent a 'flash' of the 'expanded' default button displaying when a replaceable block is deselected.

SiobhyB committed:
Tweak code comment for clarity

SiobhyB committed:
Merge branch 'trunk' into rnmobile/update/default-block-inserter-button

SiobhyB committed:
Remove redundant code causing test failures

SiobhyB committed:
Revert "Remove redundant code causing test failures"

This reverts commit ac5607ee1cfc237dcf79a3c09a5292edff10721c.

SiobhyB commented:
For posterity, noting that I've converted this PR back to a draft to reflect the fact it's a work in progress following the discovery of a blocking bug.

fluiddot committed:
Revert "Turn 'isDefaultView' into 'global' editor setting"

This reverts commit fd73fa68479de0edd6433af14697a8c2896befdf.

fluiddot committed:
Revert "Prevent unwanted 'flash' of 'expanded' button"

This reverts commit db5d32e37984bf77fba7250f0dc0179b9715add3.

# Conflicts:
#	packages/block-editor/src/components/inserter/index.native.js

fluiddot committed:
Preserve isDefaultView value while inserter is open

fluiddot committed:
Rename isInserterOpen state value

fluiddot commented:
@SiobhyB I've just pushed a fix in https://github.com/WordPress/gutenberg/pull/39726/commits/bc9b4ea72674bbe43d5070deb0239830897374c9 for the visual glitch outlined in https://github.com/WordPress/gutenberg/pull/39726#issuecomment-1091752804. This has implied reverting your latest changes related to using the global editor settings (revert commits: https://github.com/WordPress/gutenberg/pull/39726/commits/24d929482b74c93eda78de3b8c66698d1c507648 https://github.com/WordPress/gutenberg/pull/39726/commits/adf4173cc02b5c09b713cab86243cddb355593d7). If you agree, we can set the PR back to ready for review and I can proceed with reviewing it, wdyt?

fluiddot committed:
Apply BEM naming in header toolbar styles

SiobhyB commented:
@fluiddot, that sounds perfect, thank you 🙇‍♀️

fluiddot committed:
Rename default view to expanded mode

fluiddot committed:
Use hasSelectedBlock selector

fluiddot committed:
Fix header toolbar height in layout component

fluiddot commented on packages/edit-post/src/components/layout/index.native.js:
```diff
@@ -115,7 +115,7 @@ class Layout extends Component {
 		const marginBottom =
 			Platform.OS === 'android' && ! isHtmlView
-				? headerToolbarStyles.container.height
+				? headerToolbarStyles[ 'header-toolbar__container' ].height
```
This value takes the default height of the header toolbar, so it doesn't take into account the case when the toolbar is displayed in the expanded mode. As far as I tested, I didn't identify any visual issue regarding this logic, so looks safe to only use that value.

In the future, it would be great to update this calculation, and instead of accessing the height from the styles, retrieve it from the `onLayout` event.

fluiddot reviewed (approved):
LGTM 🎊 !

_Tested on iPhone 11 (iOS 15.4.1) and Samsung Galaxy S20 FE 5G (Android 11)._

fluiddot committed:
Add style mock

fluiddot commented:
PR check `End-to-End Tests / Admin - 2` is failing consistently, I checked [the latest commit from `trunk` branch](https://github.com/WordPress/gutenberg/commit/4ed8a414d6314932046de99376ab49c98e129eb9), and looks like the same job failed. Probably the failure is not caused by these changes but we should verify this issue before merging.

fluiddot commented:
> PR check `End-to-End Tests / Admin - 2` is failing consistently, I checked [the latest commit from `trunk` branch](https://github.com/WordPress/gutenberg/commit/4ed8a414d6314932046de99376ab49c98e129eb9), and looks like the same job failed. Probably the failure is not caused by these changes but we should verify this issue before merging.

I'm going to update the PR with `trunk` as it looks like the PR check that is failing is already fixed per this commit: https://github.com/WordPress/gutenberg/commit/eb61ff9aa493ed6d9dabc4926a516cb65b44689e.

fluiddot committed:
Merge branch 'trunk' into rnmobile/update/default-block-inserter-button
